pkg/sql/opt/optbuilder/mutation_builder_unique.go

```diff
@@ -235,27 +260,40 @@ func (h *uniqueCheckHelper) buildInsertionCheck() memo.UniqueChecksItem {
 	}
 	semiJoinFilters = append(semiJoinFilters, f.ConstructFiltersItem(pkFilter))
 
-	semiJoin := f.ConstructSemiJoin(checkInput, scanScope.expr, semiJoinFilters, memo.EmptyJoinPrivate)
+	semiJoin := f.ConstructSemiJoin(withScanScope.expr, scanScope.expr, semiJoinFilters, memo.EmptyJoinPrivate)
+
+	// Collect the key columns that will be shown in the error message if there
+	// is a duplicate key violation resulting from this uniqueness check.
+	keyCols := make(opt.ColList, 0, h.uniqueOrdinals.Len())
+	for i, ok := h.uniqueOrdinals.Next(0); ok; i, ok = h.uniqueOrdinals.Next(i + 1) {
+		keyCols = append(keyCols, withScanScope.cols[i].id)
+	}
 
 	return f.ConstructUniqueChecksItem(semiJoin, &memo.UniqueChecksItemPrivate{
 		Table:        h.mb.tabID,
 		CheckOrdinal: h.uniqueOrdinal,
-		// uniqueOrdinals is always a prefix of uniqueAndPrimaryKeyOrdinals, which
-		// maps 1-to-1 to the columns in withScanCols. The remaining columns are
-		// primary key columns and should not be included in the KeyCols.
-		KeyCols: withScanCols[:len(h.uniqueOrdinals)],
-		OpName:  h.mb.opName,
+		KeyCols:      keyCols,
+		OpName:       h.mb.opName,
 	})
 }
 
-// buildTableScan builds a Scan of the table.
-func (h *uniqueCheckHelper) buildTableScan() (outScope *scope, tabMeta *opt.TableMeta) {
-	tabMeta = h.mb.b.addTable(h.mb.tab, tree.NewUnqualifiedTableName(h.mb.tab.Name()))
+// buildTableScan builds a Scan of the table. The ordinals of the columns
+// scanned are also returned.
+func (h *uniqueCheckHelper) buildTableScan() (outScope *scope, ordinals []int) {
+	tabMeta := h.mb.b.addTable(h.mb.tab, tree.NewUnqualifiedTableName(h.mb.tab.Name()))
+	// TODO(mgartner): Should mutation columns be included here? Does uniqueness
+	// need to hold true for write-only columns?
```

Comment:
I wasn't sure how to reason about write-only columns in this case. Any insight would be helpful!